README.md
# FreeTDS compiled by Navigating Cancer SRE

This is an unmodified build of FreeTDS 1.00.54, with certain features
enabled with compile flags. Mainly OpenSSL (which is not included in 
Debian/Ubuntu's apt repos, due to licensing issues mentioned below.)

This build sets a default TDS protocol version to 7.3 (SQL Server 2008). This can be overridden
via freetds.conf or by TinyTDS/any other client, so is merely a best fallback default.

See http://www.freetds.org/userguide/choosingtdsprotocol.htm for more information on TDS protocol versions.

### Compile Flags in this Package

* *`--with-openssl`* This is compiled with OpenSSL support. Requires OpenSSL > 1.0.1.
* *`--with-tdsver=7.3`* (As mentioned above)
* *`--enable-msdblib`* (Enable Microsoft behavior in the DB-Library API where it diverges from Sybase's.)
* *`--enable-krb5`* (Enable Kerberos support. With Kerberos you can connect to server using your stored Kerberos ticket.)
* *`--disable-debug`* (Removes debug symbols for performance)


### Debian Control Data
```
Source: freetds
Section: libs
Priority: extra
Maintainer: Navigating Cancer SRE Team <sre@navigatingcancer.com>
Build-Depends: debhelper (>=9), dh-autoreconf, openssl (>=1.0.1)
Standards-Version: 3.9.6
Homepage: https://github.com/NavigatingCancer/freetds

Package: freetds
Architecture: any
Multi-Arch: foreign
Depends: ${misc:Depends}, ${shlibs:Depends}
Description: FreeTDS 1.00.54 SQL Server Libraries
 Aug 2017 by Arthur Kepler @ Navigating Cancer SRE.
 Compiled with OpenSSL, Kerberos5, and TDSver=7.3 flags.
 Version 1.00.54 - FreeTDS Stable
 Package generated by debmake
```
## Legal

FreeTDS is licensed under the GNU GPLv2. See [COPYING](COPYING).

The OpenSSL license requires mentioning of this:

This product includes software developed by the OpenSSL Project for use in the OpenSSL Toolkit (http://www.openssl.org/).




freetds-1.00.54/ChangeLog
Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sun Jun 5 00:20:06 2016 +0100

    Add some missing include to fisql utility

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Wed Aug 9 10:44:52 2017 +0100

    Fix a compatibility issue caused but _res macro definition

    Some public header defines _res macro (resolv.h header on
    some Linux).
    For compatibility reasons in sybdb.h (which is installed)
    we just undefine the offending macro.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Wed Aug 9 10:42:27 2017 +0100

    replacements: Avoid macro redefinitions

    Some public headers already redefine strlcpy and strlcat.
    As we decided to use our internal implementation undefine
    the old macro to make sure we use our versions.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Tue Aug 8 09:41:23 2017 +0100

    odbc: Check RPC name parsing

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Tue Aug 8 09:41:03 2017 +0100

    odbc: Fix parsing of RPC name

    RPC name was not parsed correctly causing invalid calls.
    This affected also SQLTables which try to call an internal
    store procedure using catalog name.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Wed Aug 2 21:11:22 2017 +0100

    odbc: fix SQLTables query build if SQL_NTS are used

    In the case SQL_NTS where used for schema and table and catalog
    was a pattern string (containing wildcards) instead of calling
    sp_tableswc the sp_tables was called causing invalid results.

    This fixes issues #136.

Author: Lars Kanis <larskanis@gmail.com>
Date:   Wed Jul 5 18:05:59 2017 +0200

    Fix out-of-tree build on MINGW.

Author: Ondřej Hošek <hosek@edvlab.tuwien.ac.at>
Date:   Sat Jun 17 20:52:49 2017 +0200

    dblib: dbsprhead: tdsdump_log buffer address only

    The buffer passed to dbsprhead might be uninitialized; printing it as
    a string doesn't make much sense. Output its address (using "%p")
    instead.

Author: Ondřej Hošek <hosek@edvlab.tuwien.ac.at>
Date:   Sat Jun 17 20:49:09 2017 +0200

    dblib: dbsprhead: adjust buf_len by name length

    After copying the column name into the buffer in dbsprhead, reduce
    buf_len, which is used to guard against writing outside the buffer, by
    the length of the column name.

Author: Ondřej Hošek <hosek@edvlab.tuwien.ac.at>
Date:   Sat Jun 17 19:51:04 2017 +0200

    tds: Escape left brace in regex in encodings.pl

    Since Perl v5.22.0, unescaped left braces ('{') in regexes are
    deprecated; since Perl v5.26.0, they are forbidden in most cases.
    Escape such a case in encodings.pl using a backslash.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sun Jun 11 09:04:41 2017 +0100

    ctlib: fix type cast

    The argument is a CS_BOOL, not a char. This worked only on little
    endian platforms.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Thu Jun 8 13:53:07 2017 +0100

    support Linux system < 2.6.27 having eventfd

    Before this version eventfd didn't accept and have flags support.
    Try to create the eventfd without flags and then set later.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Tue Jun 6 21:59:57 2017 +0100

    pool: Avoid crash if password setting is missing

    Test on user.c was causing a crash due to NULL dereference.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Tue Jun 6 08:04:14 2017 +0100

    Fix ct_options for CS_OPT_ARITHIGNORE and CS_OPT_ARITHABORT

    The two options required just normal boolean processing.

    This problem was reported by Matt Meissner.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Wed May 24 14:46:43 2017 +0100

    Fix bulk copy setting

    By default bulk copy is enabled but the usage of bulk_copy
    field in TDSLOGIN where a bit weird.
    Make clear and document that bulk_copy != 0 means enabled.
    Also fix a mismatch between BCP_SETL and bcp_getl.
    This fix bug #119 ("BCP_SETL() and bcp_getl() have inconsistant API").

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Wed May 17 15:28:56 2017 +0100

    Fix compile if TLS is disabled

    tds_ssl_deinit have to be defined including the proper header.
    This as if TLS is disabled the function is defined in the header.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Tue May 2 23:26:24 2017 +0100

    Fix very big login packet using MSSQL

    If login packet is really big to split into multiple TDS packets
    and is encrypted MSSQL only encrypt the first TDS packet, the others
    are not encrypted.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sat Apr 29 22:57:00 2017 +0100

    Prevent invalid record parameter

    This can happen if invalid data are sent.
    For instance if server send some parameter using PARAMFMT followed
    by old style PARAM. You potentially get a result which is half
    composed by a row and half by separate data.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sat Apr 29 22:27:47 2017 +0100

    Assure SELECT @@PID query returns a normal row

    Invalid server response can cause a different row type
    to be returned causing a NULL pointer dereference in tds_set_spid.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sat Apr 29 21:14:46 2017 +0100

    Add an invalid data type to avoid not initialized pointer to functions

    Always initialize TDSCOLUMN->funcs to avoid NULL pointer dereferences.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sat Apr 29 18:53:18 2017 +0100

    Do not send clear text password if encryption login is required on Sybase

    Password should just be sent encrypted, no reason to send also
    not encrypted.

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sat Apr 15 15:49:12 2017 +0100

    Check server is not returning negative values for column sizes

Author: Frediano Ziglio <freddy77@gmail.com>
Date:   Sun Apr 2 15:28:07 2017 +0100

    Handle errors getting datatype meta information

    This prevent server to send wrong encoded types


freetds-1.00.54/src/tds/tds_willconvert.h
/*** Please do not edit this file!                                       ***/
/*** It was generated with 'perl tds_willconvert.pl > tds_willconvert.h' ***/
/*** It is much easier to edit the __DATA__ table than this file.        ***/
/***                                                                     ***/
/***                                                          Thank you. ***/

static const TDS_TINYINT type2category[256] = {
	0, /* 0 */
	0, /* 1 */
	0, /* 2 */
	0, /* 3 */
	0, /* 4 */
	0, /* 5 */
	0, /* 6 */
	0, /* 7 */
	0, /* 8 */
	0, /* 9 */
	0, /* 10 */
	0, /* 11 */
	0, /* 12 */
	0, /* 13 */
	0, /* 14 */
	0, /* 15 */
	0, /* 16 */
	0, /* 17 */
	0, /* 18 */
	0, /* 19 */
	0, /* 20 */
	0, /* 21 */
	0, /* 22 */
	0, /* 23 */
	0, /* 24 */
	0, /* 25 */
	0, /* 26 */
	0, /* 27 */
	0, /* 28 */
	0, /* 29 */
	0, /* 30 */
	0, /* SYBVOID */
	0, /* 32 */
	0, /* 33 */
	5, /* SYBIMAGE */
	7, /* SYBTEXT */
	6, /* SYBBLOB */
	5, /* SYBVARBINARY */
	0, /* SYBINTN */
	7, /* SYBVARCHAR */
	4, /* SYBMSDATE */
	4, /* SYBMSTIME */
	4, /* SYBMSDATETIME2 */
	4, /* SYBMSDATETIMEOFFSET */
	0, /* 44 */
	5, /* SYBBINARY */
	0, /* SYBINTERVAL */
	7, /* SYBCHAR */
	2, /* SYBINT1 */
	4, /* SYBDATE */
	3, /* SYBBIT */
	4, /* SYBTIME */
	2, /* SYBINT2 */
	0, /* 53 */
	0, /* 54 */
	0, /* 55 */
	2, /* SYBINT4 */
	0, /* 57 */
	4, /* SYBDATETIME4 */
	2, /* SYBREAL */
	2, /* SYBMONEY */
	4, /* SYBDATETIME */
	2, /* SYBFLT8 */
	0, /* 63 */
	2, /* SYBUINT1 */
	2, /* SYBUINT2 */
	2, /* SYBUINT4 */
	2, /* SYBUINT8 */
	0, /* SYBUINTN */
	0, /* 69 */
	0, /* 70 */
	0, /* 71 */
	0, /* 72 */
	0, /* 73 */
	0, /* 74 */
	0, /* 75 */
	0, /* 76 */
	0, /* 77 */
	0, /* 78 */
	0, /* 79 */
	0, /* 80 */
	0, /* 81 */
	0, /* 82 */
	0, /* 83 */
	0, /* 84 */
	0, /* 85 */
	0, /* 86 */
	0, /* 87 */
	0, /* 88 */
	0, /* 89 */
	0, /* 90 */
	0, /* 91 */
	0, /* 92 */
	0, /* 93 */
	0, /* 94 */
	0, /* 95 */
	0, /* 96 */
	0, /* 97 */
	0, /* SYBVARIANT */
	0, /* SYBNTEXT */
	0, /* 100 */
	0, /* 101 */
	0, /* 102 */
	1, /* SYBSENSITIVITY */
	3, /* SYBBOUNDARY */
	0, /* 105 */
	3, /* SYBDECIMAL */
	0, /* 107 */
	3, /* SYBNUMERIC */
	0, /* SYBFLTN */
	0, /* SYBMONEYN */
	0, /* SYBDATETIMN */
	0, /* 112 */
	0, /* 113 */
	0, /* 114 */
	0, /* 115 */
	0, /* 116 */
	0, /* 117 */
	0, /* 118 */
	0, /* 119 */
	0, /* 120 */
	0, /* 121 */
	2, /* SYBMONEY4 */
	0, /* SYBDATEN */
	0, /* 124 */
	0, /* 125 */
	0, /* 126 */
	2, /* SYBINT8 */
	0, /* 128 */
	0, /* 129 */
	0, /* 130 */
	0, /* 131 */
	0, /* 132 */
	0, /* 133 */
	0, /* 134 */
	0, /* 135 */
	0, /* 136 */
	0, /* 137 */
	0, /* 138 */
	0, /* 139 */
	0, /* 140 */
	0, /* 141 */
	0, /* 142 */
	0, /* 143 */
	0, /* 144 */
	0, /* 145 */
	0, /* 146 */
	0, /* SYBTIMEN */
	0, /* 148 */
	0, /* 149 */
	0, /* 150 */
	0, /* 151 */
	0, /* 152 */
	0, /* 153 */
	0, /* 154 */
	0, /* 155 */
	0, /* 156 */
	0, /* 157 */
	0, /* 158 */
	0, /* 159 */
	0, /* 160 */
	0, /* 161 */
	0, /* 162 */
	0, /* SYBXML */
	0, /* 164 */
	5, /* XSYBVARBINARY */
	0, /* 166 */
	7, /* XSYBVARCHAR */
	0, /* 168 */
	0, /* 169 */
	0, /* 170 */
	0, /* 171 */
	0, /* 172 */
	5, /* XSYBBINARY */
	0, /* SYBUNITEXT */
	7, /* SYBLONGCHAR */
	0, /* SYBSINT1 */
	0, /* 177 */
	0, /* 178 */
	0, /* 179 */
	0, /* 180 */
	0, /* 181 */
	0, /* 182 */
	0, /* 183 */
	0, /* 184 */
	0, /* 185 */
	0, /* 186 */
	4, /* SYB5BIGDATETIME */
	4, /* SYB5BIGTIME */
	0, /* 189 */
	0, /* 190 */
	0, /* SYB5INT8 */
	0, /* 192 */
	0, /* 193 */
	0, /* 194 */
	0, /* 195 */
	0, /* 196 */
	0, /* 197 */
	0, /* 198 */
	0, /* 199 */
	0, /* 200 */
	0, /* 201 */
	0, /* 202 */
	0, /* 203 */
	0, /* 204 */
	0, /* 205 */
	0, /* 206 */
	0, /* 207 */
	0, /* 208 */
	0, /* 209 */
	0, /* 210 */
	0, /* 211 */
	0, /* 212 */
	0, /* 213 */
	0, /* 214 */
	0, /* 215 */
	0, /* 216 */
	0, /* 217 */
	0, /* 218 */
	0, /* 219 */
	0, /* 220 */
	0, /* 221 */
	0, /* 222 */
	0, /* 223 */
	0, /* 224 */
	5, /* SYBLONGBINARY */
	0, /* 226 */
	0, /* 227 */
	0, /* 228 */
	0, /* 229 */
	0, /* 230 */
	0, /* XSYBNVARCHAR */
	0, /* 232 */
	0, /* 233 */
	0, /* 234 */
	0, /* 235 */
	0, /* 236 */
	0, /* 237 */
	0, /* 238 */
	0, /* XSYBNCHAR */
	0, /* SYBMSUDT */
	0, /* SYBMSXML */
	0, /* 242 */
	0, /* 243 */
	0, /* 244 */
	0, /* 245 */
	0, /* 246 */
	0, /* 247 */
	0, /* 248 */
	0, /* 249 */
	0, /* 250 */
	0, /* 251 */
	0, /* 252 */
	0, /* 253 */
	0, /* 254 */
	0, /* 255 */
};

static const TDS_UINT category_conversion[] = {
	0,
	0,	/* SYBSENSITIVITY */
	172,	/* SYBINT1 */
	172,	/* SYBBIT */
	176,	/* SYBMSDATE */
	164,	/* SYBIMAGE */
	224,	/* SYBUNIQUE */
	252,	/* SYBTEXT */
};
